### PR TITLE
Added Multi company support

### PR DESCRIPTION
**Pull request by Martronic-SA**
Could you check this?
I've added company_ids and currency_id in MIS report instance.
For the moment, the rate calcuation is not implemented (will be soon).

### CODE REVIEW

File: mis_builder/models/aep.py [hunk 1]
```diff
@@ -67,9 +67,12 @@ class AccountingExpressionProcessor(object):
                          r"(?P<accounts>_[a-zA-Z0-9]+|\[.*?\])"
                          r"(?P<domain>\[.*?\])?")
 
-    def __init__(self, company):
-        self.company = company
-        self.dp = company.currency_id.decimal_places
+    def __init__(self, company_ids=None, currency_id=None):
```
Comment (sbidoul) [inline mis_builder/models/aep.py]: Can you name that `companies` instead of `company_ids`.

File: mis_builder/models/aep.py [hunk 2]
```diff
@@ -255,9 +258,9 @@ def do_queries(self, date_from, date_to,
         This method must be executed after done_parsing().
         """
         if not aml_model:
-            aml_model = self.company.env['account.move.line']
+            aml_model = self.company_ids[0].env['account.move.line']
         else:
-            aml_model = self.company.env[aml_model]
+            aml_model = self.company_ids[0].env[aml_model]
```
Comment (sbidoul) [inline mis_builder/models/aep.py]: `self.companies.env[aml_model]` : `[0]` is not needed

File: mis_builder/models/aep.py [hunk 3]
```diff
@@ -67,9 +67,12 @@ class AccountingExpressionProcessor(object):
...
 
-    def __init__(self, company):
-        self.company = company
-        self.dp = company.currency_id.decimal_places
+    def __init__(self, company_ids=None, currency_id=None):
+        self.company_ids = company_ids and company_ids or \
+            [self.env.user.company_id]
+        self.currency_id = currency_id and currency_id or \
+            self.env.user.company_id.currency_id
+        self.dp = self.currency_id.decimal_places
```
Comment (sbidoul) [inline mis_builder/models/aep.py]: Could you keep the currency stuff in separate commits? So we different commit(s) for multi company and multi currency and a clean commit history.

File: mis_builder/models/aep.py [hunk 4]
```diff
@@ -238,7 +241,7 @@ def get_aml_domain_for_dates(self, date_from, date_to,
         elif mode == self.MODE_UNALLOCATED:
             date_from_date = fields.Date.from_string(date_from)
             fy_date_from = \
-                self.company.\
+                self.company_ids[0].\
```
Comment (sbidoul) [inline mis_builder/models/aep.py]: Hm. Here it's a special treatment for the first company. This could lead to surprises. 

Could you extract the `fy_date_from` computation in a separate method and add a check (raise UserError) that all companies have the same `fy_date_from`? Later we can add logic to handle companies with different fiscal years.

File: mis_builder/models/mis_report_instance.py [hunk 5]
```diff
@@ -348,10 +348,14 @@ def _default_company(self):
...
-    company_id = fields.Many2one(comodel_name='res.company',
-                                 string='Company',
-                                 default=_default_company,
-                                 required=True)
+    company_ids = fields.Many2many(comodel_name='res.company',
+        string='Company',
+        help='Select companies for which data will  be searched. \
+            User\'s company if empty.',
+        default=_default_company_ids,
+        required=False)
```
Comment (sbidoul) [inline mis_builder/models/mis_report_instance.py]: Could you use this style for new/modified fields?

```python
company_ids = fields.Many2many(
        comodel_name='res.company',
        string='Company',
        help='Select companies for which data will  be searched. '
                 'User\'s company if empty.',
        default=_default_company_ids,
        required=False,
)
```

Comment (sbidoul) [reply mis_builder/models/mis_report_instance.py]: A migration script is needed.

File: mis_builder/models/mis_report_instance.py [hunk 6]
```diff
@@ -348,10 +348,14 @@ def _default_company(self):
...
+    currency_id = fields.Many2one('res.currency', 'Currency',
+        required=False)
```
Comment (sbidoul) [inline mis_builder/models/mis_report_instance.py]: Could you please do multi currency in another PR?

File: mis_builder/models/mis_report_instance.py [hunk 7]
```diff
@@ -320,9 +320,9 @@ def _compute_pivot_date(self):
                 record.pivot_date = fields.Date.context_today(record)
 
     @api.model
-    def _default_company(self):
-        return self.env['res.company'].\
-            _company_default_get('mis.report.instance')
+    def _default_company_ids(self):
+        return [(6,0,[self.env['res.company'].\
+            _company_default_get('mis.report.instance').id])]
```
Comment (sbidoul) [inline mis_builder/models/mis_report_instance.py]: flake8 will complain here

Comment (sbidoul) [reply mis_builder/models/mis_report_instance.py]: I'm wondering if `return _company_default_get('mis.report.instance')` would work too?

**Review by sbidoul — CHANGES_REQUESTED** [review]
Thanks for this contrib!

**Comment by sbidoul** [comment]
Can you add a new test for multi-company AEP?

File: mis_builder/tests/test_aep.py [hunk 8]
```diff
@@ -214,15 +215,15 @@ def test_aep_convenience_methods(self):
...
+            self.company_ids,
             time.strftime('%Y') + '-03-31',
             'posted')
         self.assertEquals(end, {
             self.account_ar.id: (900, 0),
             self.account_in.id: (0, 800),
         })
         unallocated = AEP.get_unallocated_pl(
-            self.company,
+            self.company_ids,
```
Comment (sbidoul) [inline mis_builder/tests/test_aep.py]: `test_aep.py` should be unchanged: existing tests in this file should continue to work without change.

File: mis_builder/models/mis_report_instance.py (hunk 7 above)
Comment (Martronic-SA) [reply mis_builder/models/mis_report_instance.py]: > I'm wondering if return _company_default_get('mis.report.instance') would work too?

That's what I was using first, but there were a json complaint about res.company not serialisable.
Why flake8 will complain?

File: mis_builder/tests/test_aep.py (hunk 8 above)
Comment (Martronic-SA) [reply mis_builder/tests/test_aep.py]: So you the the AEP __init__ method shoud accept company and companies as first argument and I shoud add check whether theare are one or many companies passed?

File: mis_builder/models/aep.py (hunk 4 above)
Comment (Martronic-SA) [reply mis_builder/models/aep.py]: I can see some border effects when company is located in Australia with GMT +12, but this can be fixed.
However don't you think that for this particular case (at least) the main company (or user.company_id) is useful?

File: mis_builder/models/mis_report_instance.py (hunk 7 above)
Comment (sbidoul) [reply mis_builder/models/mis_report_instance.py]: flake8 requires a space after commas

File: mis_builder/tests/test_aep.py (hunk 8 above)
Comment (sbidoul) [reply mis_builder/tests/test_aep.py]: What I mean is that for Odoo, a recordset of 1 is still a recordset, and `create` returns a recordset of 1. So these tests can remain unmodified and work normally even if `AEP.__init__` accept companies as argument.

File: mis_builder/models/aep.py (hunk 4 above)
Comment (sbidoul) [reply mis_builder/models/aep.py]: It could be useful to have a main company but not in AEP. Perhaps in mis_report_instance. But even that I'm not sure and I certainly have not studied the impact of having different fiscal years on the computation of consolidated initial balances... Until that is clarified I prefer to raise an error if the various companies have different fiscal year dates.

**Comment by Martronic-SA** [comment]
Even if currency support is in a separate request, we still need to choose a main currency for total amounts calculation as account.move.lines might be in different currencies.
Do you see a way to output the main currency in the report?
What do you think about having a cell of type "Monetary" that would format the number as Monetary field according to the main currency in report? or a new checkbox in the style for "Monetary" ?

EDITED : This [PR](https://github.com/Martronic-SA/account-financial-reporting/pull/5)

**Comment by sbidoul** [comment]
Regarding this PR, a few things remain todo before merging;

- a test for multi-company
- a few flak8 errors: https://travis-ci.org/acsone/account-financial-reporting/jobs/245282951#L546-L548
- a migration script for mis.report.instance.company_id -> company_ids

Regarding multi-currency I would not worry too much with the display issue: we currently do it with a suffix/prefix in the styles, it does the job. Doing the currency conversion in an efficient manner is the key problem to resolve.

**Comment by sbidoul** [comment]
Please please do the currency conversion in another pull request so this one remains reasonably small and easy to review.

**Comment by Martronic-SA** [comment]
I've added a basic currency conversion of amounts to the main currency.
(For now, the main currency is the user's company currency.)
This will at least avoid having amounts in different currencies mixed together.
This does not support choosing currency rates at a particular date..

File: mis_builder/models/aep.py [hunk 9]
```diff
@@ -283,14 +284,20 @@ def do_queries(self, date_from, date_to,
...
                 if mode in (self.MODE_INITIAL, self.MODE_UNALLOCATED) and \
                         float_is_zero(debit-credit,
-                                      precision_rounding=self.dp):
+                            precision_rounding=\
+                            company.currency_id.decimal_places):
                     # in initial mode, ignore accounts with 0 balance
                     continue
-                self._data[key][acc['account_id'][0]] = (debit, credit)
+                if company.currency_id != self.currency_id:
+                    rate = self.currency_id.rate / company.currency_id.rate
```
Comment (Martronic-SA) [inline mis_builder/models/aep.py]: The basic implementation is quite simple.
Get the account for the line, get its company. If account company currency is not the same as APE currency, do the conversion (using current rate for now)

File: mis_builder/models/aep.py [hunk 10]
```diff
@@ -281,14 +284,20 @@ def do_queries(self, date_from, date_to,
                                         ['debit', 'credit', 'account_id'],
                                         ['account_id'])
             for acc in accs:
+                company = account_model.browse(acc['account_id'][0]).company_id
```
Comment (sbidoul) [inline mis_builder/models/aep.py]: This browse inside the loop is bad for performance.

I suggest:
1. fetching and computing currency rates by company at the beginning of the method, outside of the loops, store that in a dictionary
2. adding company_id in the read_group

Comment (Martronic-SA) [reply mis_builder/models/aep.py]: Yes good idea!

Comment (Martronic-SA) [reply mis_builder/models/aep.py]: What do you think about the precision rounding in float_is_zero method ? should it be company currency rounding or AEP currency rounding?

File: mis_builder/models/aep.py [hunk 11]
```diff
@@ -67,9 +67,11 @@ class AccountingExpressionProcessor(object):
                          r"(?P<accounts>_[a-zA-Z0-9]+|\[.*?\])"
                          r"(?P<domain>\[.*?\])?")
 
-    def __init__(self, company):
-        self.company = company
-        self.dp = company.currency_id.decimal_places
+    def __init__(self, companies, currency_id=None):
+        self.companies = companies
+        self.currency_id = currency_id and currency_id or \
+            self.companies.env.user.company_id.currency_id
```
Comment (sbidoul) [inline mis_builder/models/aep.py]: here, raise UserError if currency_id is not set and companies don't have all the same currency

File: mis_builder/models/aep.py (hunk 10 above)
Comment (sbidoul) [reply mis_builder/models/aep.py]: Use self.dp, which should come from the target currency.

Comment (Martronic-SA) [reply mis_builder/models/aep.py]: But if target currency is CHF which is 0.5 rounding and we're getting amounts in EUR which rounding is 0.1, won't we miss some amounts? That could be problematic especially if there are many amounts with 0.1 or 0.2 ?

File: mis_builder/models/aep.py (hunk 11 above)
Comment (Martronic-SA) [reply mis_builder/models/aep.py]: Ok, but we might also consider that user wants to see the amounts in his own currency by default. But then we would need the Monetary cell as if not set, the traget currency depends on user

Comment (sbidoul) [reply mis_builder/models/aep.py]: In that case the user will specify the currency explicitly in mis_report_instance.
AEP is a low level API, and I want it to have as little implicit behaviour as possible.

Comment (Martronic-SA) [reply mis_builder/models/aep.py]: Ok, that's a good reason.

File: mis_builder/models/aep.py (hunk 10 above)
Comment (sbidoul) [reply mis_builder/models/aep.py]: Hm. Indeed this float_compare occurs before currency conversion, so you are probably right, yes.

**Comment by Martronic-SA** [comment]
TODO:
**Testing**
test the tests, I've added an test for multi company. I've not been able to setup a testing environment and I've not been able to test those test files.
There is missing a proper test for currency conversion (set a currency rate and check the final results).
(But I've tested on a dev DB and it's working fine.)

**Migration**
- [x] Write the migration script to convert from company to company_ids in mis.report.instance

There is missing some functionalities to have this 100% usable, at least:

- [x] Add currency field in the mis.report.instance (separate PR)
- [ ] Display a number as a monetary field and format it according to the target currency. Might be done either with a new type of cell "Monetary" or a checkbox in the style "Monetary formatting".
- [ ] Add a selection field with options to tell the AEP how to choose the right date for currency rate conversion. That could be now, start/end of the report, start/end of each period or an average on the report/period, or..?

**Comment by Martronic-SA** [comment]
I think we are ready to add the currency_id field to mis report instance?
It's not a big change, I've done it in a sapearate branch here and it's working fine: 
[10.0-mis_builder-3-currency](https://github.com/Martronic-SA/account-financial-reporting/tree/10.0-mis_builder-3-currency)

**Comment by sbidoul** [comment]
Yes, now that all the multi-currency logic in AEP is in this branch, it's a small change to add curency_id in mis_report_instance in this PR.

File: mis_builder/migrations/8.0.2.0.0/pre-migration.py [hunk 12]
```diff
@@ -2,6 +2,66 @@
...
+          USING btree
+          (res_company_id);
+    """)
+    
+    cr.execute(
+        """INSERT INTO mis_report_instance_res_company_rel
+             SELECT id as mis_report_instance_id,
+                    company_id as res_company_id
+             FROM mis_report_instance;
+        """)
```
Comment (sbidoul) [inline mis_builder/migrations/8.0.2.0.0/pre-migration.py]: Hm that's a lot of code.

Better use openupgradelib for this: https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L789

Also bump version to 10.0.3.1.0 and make a different migration directory for that.

Comment (Martronic-SA) [reply mis_builder/migrations/8.0.2.0.0/pre-migration.py]: So much simpler, thank you for remind me the openupgrade

**Comment by Martronic-SA** [comment]
I think we should add company_ids to mis.report.instance.period and if not null, add the company filter to
_get_additional_move_line_filter and _get_additional_query_filter methods.
This is useful if you want to make a comparision between companies in a table with a company per column.

We might better add company_ids filter field to sub kpi filter

**Comment by sbidoul** [comment]
Thanks a lot for your work! I've too little time to finalize the review and merge now, but I keep an eye on this and will do as soon as I can.

**Comment by Martronic-SA** [comment]
Please check also this [PR](https://github.com/Martronic-SA/account-financial-reporting/pull/5)
It's the Monetary type of value which takes the currency set in report instance or user's currency to format the values automatically.

**Comment by sbidoul** [comment]
@Martronic-SA Thanks for your contributions and patience.

I'm now back working on this and I'm planning to do a release with as many new features as possible before end of september.

I'll now flesh out the tests a little bit and merge at least some parts of this multi company PR.

Could you rebase it on OCA#275 ?

**Comment by Martronic-SA** [comment]
@sbidoul I'm back from holidays.
Unfortunately I'm having trouble with rebasing. I'm not very used to git in fact.
Any advice?
Thank you

**Comment by sbidoul** [comment]
Here is a rough rebasing procedure:

- make sure you have a git remote named `acsone`: `git remote add acsone https://github.com/acsone/account-financial-reporting.git`
- `git fetch --all`
- `git rebase acsone/10.0-mis_builder-3-sbi`
- in case of conflict, edit the conflicting files to resolve the conflict (look for === in the files), then `git add` the corrected files, the `git rebase --continue`
- the `git push -f`

**Comment by sbidoul** [comment]
@Martronic-SA will you be at the OCA code sprint in Louvain-la-Neuve?

**Comment by Martronic-SA** [comment]
Thank you very much for the procedure.
I think it's ok now with the rebase.
Unfortunately I can't be there in Louvain-La-Neuve next week but I will be there remotely if possible.

**Comment by sbidoul** [comment]
@Martronic-SA could you review https://github.com/OCA/account-financial-reporting/pull/275 ?

We'll merge that first then move on with your other PRs.

**Comment by Martronic-SA** [comment]
I think everything is fine with [OCA#275](https://github.com/OCA/account-financial-reporting/pull/275). ok for me!

**Comment by sbidoul** [comment]
Can write your approval there?

**Comment by sbidoul** [comment]
I'm currently moving this PR to OCA/mis-builder.

**Comment by sbidoul** [comment]
Moved to https://github.com/OCA/mis-builder/pull/23
